extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs

```diff
@@ -0,0 +1,286 @@
+            }
+        }
+
+        private bool IsCollectionBinding(Type targetType, string blobName, out Type? elementType)
+        {
+            elementType = null;
+
+            // Edge case: These two types should be treated as a single blob binding
+            // string implements IEnumerable<char> and byte[] would pass the IsArray check
+            if (targetType == typeof(string) || targetType == typeof(byte[]))
+            {
+                return false;
+            }
+
+            if (!(targetType.IsArray || typeof(IEnumerable).IsAssignableFrom(targetType)))
+            {
+                return false;
+            }
+
+            // At this stage, we know we have a collection type binding
+            elementType = targetType.IsArray ? targetType.GetElementType() : targetType.GenericTypeArguments[0];
+
+            if (elementType == typeof(BlobContainerClient))
+            {
+                throw new InvalidOperationException("Binding to a BlobContainerClient collection is not supported.");
+            }
+
+            bool isFile = BlobIsFileRegex.IsMatch(blobName);
+
+            if (isFile && typeof(BlobBaseClient).IsAssignableFrom(elementType))
+            {
+                throw new InvalidOperationException("Binding to a blob client collection with a blob path is not supported. "
+                                                    + "Either bind to the container path, or use BlobClient instead.");
+            }
+
+            // If it's a collection binding and the blob is a file, we should treat it as a single blob binding as it
+            // end up in the `DeserializeToTargetObjectAsync` method via ToTargetTypeAsync()
+            if (isFile)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private async Task<object> BindToCollectionAsync(Type targetType, Type elementType, BlobContainerClient container, string blobPath)
+        {
+            var resultType = typeof(List<>).MakeGenericType(elementType);
+            var result = (IList)Activator.CreateInstance(resultType);
+
+            await foreach (BlobItem blobItem in container.GetBlobsAsync(prefix: blobPath).ConfigureAwait(false))
+            {
+                var element = await ToTargetTypeAsync(elementType, container, blobItem.Name);
+                result.Add(element);
+            }
+
+            if (targetType.IsArray)
+            {
+                var arrayResult = Array.CreateInstance(elementType, result.Count);
+                result.CopyTo(arrayResult, 0);
+                return arrayResult;
+            }
+
+            return result;
+        }
+
+        private async Task<object?> ToTargetTypeAsync(Type targetType, BlobContainerClient containerClient, string blobName) => targetType switch
+        {
+            Type _ when targetType == typeof(BlobContainerClient) => containerClient,
```

Comment:
do we wanna create separate converters for each target type? This can definitely be done in another PR too since this one's already so big

Reply:
We have an issue tracking this as an post-GA item